core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -572,6 +556,19 @@ public static boolean assertNotClusterStateUpdateThread(String reason) {
         return true;
     }
 
+    /** asserts that the current stack trace does <b>NOT</b> invlove a cluster state applier */
+    private static boolean assertNotCalledFromClusterStateApplier(String reason) {
```

Comment:
I wonder if `assertNotCalledFromClusterStateApplier` should also add checks for the call to `clusterSettings.applySettings` that is done before the state is updated. ZenDiscovery for example adds a settings update consumer that accesses ClusterService.state().